### PR TITLE
frontend: restore legacy test history url and redundant test result bug fix

### LINKED ISSUE #758: Test case results comparison broken old link
With the new squad release, The changes to test results comparison has changed.
I have reported handful of bugs with quoting qa report test results comparison of a given branch
These kind of listed links are broken now when viewed by external and internal users from Bugzilla and upstream email reports which i shared.

https://qa-reports.linaro.org/lkft/linux-stable-rc-4.4-oe/tests/ltp-syscalls-tests/fanotify10
https://qa-reports.linaro.org/lkft/linux-stable-rc-4.9-oe/tests/ltp-syscalls-tests/fanotify10
https://qa-reports.linaro.org/lkft/linux-stable-rc-5.6-oe/tests/ltp-syscalls-tests/clock_gettime03
https://qa-reports.linaro.org/lkft/linux-mainline-oe/tests/ltp-mm-tests/shmt03
https://qa-reports.linaro.org/lkft/linux-mainline-oe/tests/kselftest/net_forwarding_bridge_igmp.sh
and many more

- Naresh

### PR DESCRIPTION
**Pull request by amrohassaan**
Fixes #758

### CODE REVIEW

**Comment by amrohassaan** [comment]
@mwasilew , @chaws reason i combined restoring the test history url and the fix for the redundant test results in build view in a single patch is that they are both related and latter depends on the old history url. That test results redundant bug was introduced in my original pr that added the test details page.

File: squad/frontend/templates/squad/tests.jinja2 [hunk 1]
```diff
@@ -29,11 +29,11 @@
         </thead>
         {% if results %}
             {% for test in results %}
-            {% set test_details_url = url('testrun_suite_test_details', args=[build.project.group.slug, build.project.slug, build.version, test.test_run.id, test.suite.slug.replace('/', '$'), test.name]) %}
+            {% set test_history_url = url('test_history', args=[build.project.group.slug, build.project.slug, test.name]) %}
```
Comment (mwasilew) [inline squad/frontend/templates/squad/tests.jinja2:32]: this doesn't work. I'm getting 404 when I click the link. Beside that, this link should lead to test details, not history. So you went one step too far :)

**Comment by mwasilew** [comment]
I tried fixing it but it means rewriting the whole patch. So I'll leave it to you.

File: squad/frontend/templates/squad/tests.jinja2 (hunk 1 above)
Comment (amrohassaan) [reply squad/frontend/templates/squad/tests.jinja2:32]: @mwasilew btw, this used to point to history before we added the details page, some rework needs to happen to the TestResults class then. lemme check the 404, you got that on the android project?

**Comment by amrohassaan** [comment]
@mwasilew @chaws changed back the link from test name in 'all test results' to the details page.
let me know if any 404s or broken pages still exist.

**Comment by mwasilew** [comment]
I found a test that results in 404:
cts-lkft-armeabi-v7a/armeabi-v7a.CtsLibcoreTestCases/libcore.java.nio.file.Files2Test.test_readAllLine$Path$Charset
But this is expected as it contains $ in the name. What is more worrying is that tests with / in the name also give 404:
cts-lkft-armeabi-v7a/armeabi-v7a.CtsLibcoreTestCases/libcore.java.time.zone.IcuZoneRulesProviderTest.testAllTransitions[Africa/Djibouti]
The URL shows sth like this: http://127.0.0.1:8000/android-lkft/ti-android-linux-4.14.y-aosp/build/v4.14.62-3132-g7793ddca6f56/testrun/828482/suite/cts-lkft-armeabi-v7a$armeabi-v7a.CtsLibcoreTestCases/test/Djibouti%5D/details/. Replacing / goes a bit too far and also removes parts of the name.

File: squad/frontend/tests.py [hunk 2]
```diff
@@ -19,6 +19,7 @@ class TestResult(list):
 
     def __init__(self, name, test_run=None, suite=None):
         self.name = name
+        self.short_name = name.rsplit('/', 1)[1]
```
Comment (mwasilew) [inline squad/frontend/tests.py:22]: this causes the problem when the test contains / in the name.

File: squad/frontend/tests.py [hunk 3]
```diff
@@ -123,8 +124,10 @@ def get(cls, build, page, search, per_page=50):
 
         memo = {}
         for test in tests:
-            memo.setdefault(test, {})
-            memo[test][test.test_run.environment_id] = [test.status]
+            memo.setdefault(test.full_name, {})
```
Comment (mwasilew) [inline squad/frontend/tests.py:127]: Using Test object as a key seems better. Why was this change needed?

File: squad/frontend/tests.py [hunk 4]
```diff
@@ -170,13 +174,19 @@ def tests(request, group_slug, project_slug, build_version):
...
+    if build_version and testrun and suite_slug:
+        build = get_build(project, build_version)
+        test_run = get_object_or_404(build.test_runs, pk=testrun)
+        suite_slug = suite_slug.replace('$', '/')
+        suite = get_object_or_404(project.suites, slug=suite_slug)
+        status = get_object_or_404(test_run.status, suite=suite)
+        full_test_name = "/".join([suite_slug, test_name])
+        context.update({"build": build, "status": status, "test": test_name})
+    else:
+        full_test_name = request.path.rsplit("tests/", 1)[1].replace('$', '/')
```
Comment (mwasilew) [inline squad/frontend/tests.py:189]: this is risky. If suite or test contains 'tests/' in the name it will result in incorrect name.

**Comment by mwasilew** [comment]
This looks and works OK. Let's merge it now and add redirection to test details in a separate patch.
